### PR TITLE
[stdlib] Make unsafe array initializer public

### LINKED ISSUE #45677: [SR-3087] No way to arbitrarily initialise an Array's storage
|                  |                 |
   |------------------|-----------------|
   |Previous ID       | SR-3087      |
   |Radar             | rdar://problem/21880692         |
   |Original Reporter | @karwa      |
   |Type              | Improvement    |
   |Status            | Resolved        |
   |Resolution        | Done    |
        
   <details>
  <summary>Additional Detail from JIRA</summary>

   |                  |                 |
   |------------------|-----------------|
   |Votes             | 0         |
   |Component/s       | Standard Library    |
   |Labels            | Improvement, LanguageFeatureRequest        |
   |Assignee          | @natecook1000      |
   |Priority          | Medium      |

   

   md5: 78d2891e46bb0f408db15202b1a1b736

  </details>

**relates to**:  
* [SR-3631](https://bugs.swift.org/browse/SR-3631) Introduce ContiguouslyStored protocol and kill ArraySlice type  

**Issue Description:**

The standard library does not currently expose a way to allocate an Array's storage and (unsafely) initialise its elements, as we can with UnsafeMutablePointer.

An example use-case is to implement a concurrent map. See: <https://lists.swift.org/pipermail/swift-users/Week-of-Mon-20161024/003769.html>

Even though we expose a UMP from Array, it cannot be used to unsafely initialise the Array elements.

The workaround at the moment is to allocate a UMP, initialise its elements unsafely, and then copy the contents in to an Array via the Array(Sequence) constructor, which is less than optimal.

### PR DESCRIPTION
**Pull request by natecook1000**
This is an implementation for the [re-pitch of SE-0223](https://github.com/apple/swift-evolution/pull/993).

The original underscored version of this initializer remains, with a deprecation notice. Also adds the initializer to `ContiguousArray`. Still to do:

- [x] Check inlining attributes, other annotations
- [x] Add tests for `ContiguousArray` and the non-underscored `Array` version

Resolves #45677.

### CODE REVIEW

**Comment by natecook1000** [comment]
@swift-ci Please smoke test

**Comment by natecook1000** [comment]
@swift-ci Please smoke test

**Comment by natecook1000** [comment]
@swift-ci Please smoke test

**Comment by natecook1000** [comment]
@airspeedswift I was expecting this to fail the stable ABI test, since it’s adding new API — does that only track changes and deletions?

**Comment by natecook1000** [comment]
@swift-ci Please smoke test Linux platform

**Comment by natecook1000** [comment]
@swift-ci Please smoke test Linux platform

File: stdlib/public/core/Array.swift [hunk 1]
```diff
@@ -1366,21 +1366,35 @@ extension Array {
 }
 
 extension Array {
+  @available(swift, deprecated: 5.1, renamed: "init(unsafeUninitializedCapacity:initializingWith:)")
```
Comment (jrose-apple) [inline stdlib/public/core/Array.swift]: This needs to be OS availability, unfortunately. It's not tied to `-swift-version` but to your deployment target.

File: stdlib/public/core/Array.swift [hunk 2]
```diff
@@ -1392,29 +1406,29 @@ extension Array {
   ///         elements you initialize.
   @inlinable
   public init(
```
Comment (jrose-apple) [inline stdlib/public/core/Array.swift]: …And this one *needs* OS availability.

**Comment by tkremenek** [comment]
The corresponding evolution proposal has been accepted.

**Comment by tkremenek** [comment]
@swift-ci clean test

**Comment by natecook1000** [comment]
@swift-ci Please test

File: stdlib/public/core/ContiguousArray.swift [hunk 3]
```diff
@@ -985,6 +985,43 @@ extension ContiguousArray {
...
+  ///     space for in the new array.
+  ///   - initializer: A closure that initializes elements and sets the count
+  ///     of the new array.
+  ///     - Parameters:
+  ///       - buffer: A buffer covering uninitialized memory with room for the
+  ///         specified number of of elements.
+  ///       - initializedCount: The count of initialized elements in the array,
+  ///         which begins as zero. Set `initializedCount` to the number of
+  ///         elements you initialize.
+  @inlinable
```
Comment (jrose-apple) [inline stdlib/public/core/ContiguousArray.swift]: Should this one also be `@_alwaysEmitIntoClient`?

Comment (natecook1000) [reply stdlib/public/core/ContiguousArray.swift]: Yes — fixed, thanks!

File: stdlib/public/core/Array.swift (hunk 1 above)
Comment (natecook1000) [reply stdlib/public/core/Array.swift]: Addressed these in b962048.

**Comment by natecook1000** [comment]
@swift-ci Please test

**Review by Catfish-Man — APPROVED** [review]
This looks good to me!

**Comment by natecook1000** [comment]
@swift-ci Please smoke test

**Pull request merged**